### PR TITLE
refactor: fold per-target python version into base rules

### PR DESCRIPTION
**Pull request by rickeylev**
Today, specifying the Python version for a target requires using the version-aware
rules in `transition.bzl` (or the generated equivalents bound to a specific Python
version). With the rules rewritten in Bazel, that functionality can be moved into
the base rules themselves. Moving the logic into the base rules simplifies the
implementation and avoids having to re-implement subtle behaviors in the wrappers
to correctly emulate the wrapped target.

For backwards compatibility, the symbols in `transition.bzl` are left as aliases
to the underlying rules.

### CODE REVIEW

File: python/config_settings/transition.bzl [hunk 1]
```diff
@@ -14,266 +14,15 @@
...
-
-def py_test(name, python_version, **kwargs):
-    return _py_rule(_py_test, _transition_py_test, name, python_version, **kwargs)
 
-def _strip_suffix(s, suffix):
-    if s.endswith(suffix):
-        return s[:-len(suffix)]
-    else:
-        return s
+py_binary = _py_binary
```
Comment (aignas) [inline python/config_settings/transition.bzl:27]: Very small nit, we could create a macro that populate the deprecation message, that way users dont need to read our chngelog, which they might not do.

File: python/config_settings/transition.bzl [hunk 2]
```diff
@@ -14,266 +14,15 @@
...
-    if "shard_count" in kwargs:
-        common_attrs["shard_count"] = kwargs.pop("shard_count")
-    if "local" in kwargs:
-        common_attrs["local"] = kwargs.pop("local")
-
-    # Binary-specific extra attributes.
-    if "output_licenses" in kwargs:
-        common_attrs["output_licenses"] = kwargs.pop("output_licenses")
-
-    rule_impl(
```
Comment (aignas) [inline python/config_settings/transition.bzl:242]: Does this mean that we will have fewer targets? The way I read the code is that the new version does not add this extra private target.

**Review by aignas — APPROVED** [review]

File: python/config_settings/transition.bzl (hunk 2 above)
Comment (rickeylev) [reply python/config_settings/transition.bzl:242]: Correct.

File: python/config_settings/transition.bzl (hunk 1 above)
Comment (rickeylev) [reply python/config_settings/transition.bzl]: Good idea. Done

**Pull request merged**